packages/entities/src/lib/update.mutation.ts

```diff
@@ -174,3 +175,55 @@ export function upsertEntities<
     return addEntities(newEntities, options)(newState, context) as S;
   };
 }
+
+/**
+ *
+ * Merge entities that exists, add those who don't
+ * Make sure all entities have an id
+ *
+ * @example
+ *
+ * // single entity
+ * store.update(upsertEntities({ id: 1, completed: true }))
+ *
+ * // or multiple entities
+ * store.update(upsertEntities([{ id: 1, completed: true }, { id: 2, completed: true }]))
+ *
+ * // or using a custom ref
+ * store.update(upsertEntities([{ id: 1, open: true }], { ref: UIEntitiesRef }))
+ *
+ */
+export function upsertEntities<
+  S extends EntitiesState<Ref>,
+  Ref extends EntitiesRef = DefaultEntitiesRef
+>(
+  entities: OrArray<Partial<getEntityType<S, Ref>>>,
+  options: AddEntitiesOptions & BaseEntityOptions<Ref> = {}
+): Reducer<S> {
+  return function (state, context) {
+    const { prepend = false, ref = defaultEntitiesRef } = options;
+    const { entitiesKey, idsKey } = ref!;
+    const idKey = getIdKey<getIdType<S, Ref>>(context, ref);
+
+    const asObject = {} as Record<getIdType<S, Ref>, getEntityType<S, Ref>>;
+    const ids = [] as getIdType<S, Ref>;
+    for (const entity of coerceArray(entities)) {
+      const id: getIdType<S, Ref> = entity[idKey];
+      // if entity exists, merge update, else add
+      if (hasEntity(id, options)(state)) {
+        asObject[id] = { ...state[entitiesKey][id], ...entity };
+      } else {
```

Comment:
Any reason you didn't leverage the existing mutations? addEntities/updateEntities?

Reply:
updateEntities has more logic because of the toModel and update function, which we would not be able to use.
I found it was the least amount of code, easy to reason about, and we only need to loop over the entities once.

but if you feel you can improve on it, feel free :)